p2panda-net/src/engine/topic_streams.rs

```diff
@@ -0,0 +1,466 @@
+        // gossip, buffering or sync.
+        self.subscribed
+            .insert(stream_id, (topic.clone(), from_network_tx));
+        self.gossip_pending.insert(topic.id(), gossip_ready_tx);
+        self.topic_to_stream
+            .entry(topic.clone())
+            .and_modify(|stream_ids| stream_ids.push(stream_id))
+            .or_insert(vec![stream_id]);
+        self.topic_id_to_stream
+            .entry(topic.id())
+            .and_modify(|stream_ids| stream_ids.push(stream_id))
+            .or_insert(vec![stream_id]);
+
+        // Hot path: If we haven't joined a gossip overlay for this topic yet, optimistically try
+        // to do it now. If this fails we should re-try sometime later using the
+        // "try_join_pending_gossips" method.
+        self.join_gossip(topic.id()).await?;
+
+        // Spawn task to establish a channel for sending messages into gossip overlay.
+        {
+            let gossip_actor_tx = self.gossip_actor_tx.clone();
+            let gossip_joined = self.gossip_joined.clone();
+            tokio::task::spawn(async move {
+                while let Some(event) = to_network_rx.recv().await {
+                    let gossip_joined = gossip_joined.read().await;
+                    if !gossip_joined.contains(&topic.id()) {
+                        // If we haven't joined the gossip yet messages will be silently dropped
+                        // here.
+                        //
+                        // For now this is fine as the user has two options:
+                        //
+                        // 1. They're combining sync with gossip. If the user stores all messages
+                        //    before sending them (which they probably always should if they care
+                        //    about consistency) sync will make sure that peers will catch up with
+                        //    this data as soon as they connect to somebody.
+                        // 2. They're don't care about consistency, but they are waiting for the
+                        //    "gossip ready" signal before they send any messages.
+                        continue;
+                    }
+
+                    let result = match event {
+                        ToNetwork::Message { bytes } => {
+                            gossip_actor_tx
+                                .send(ToGossipActor::Broadcast {
+                                    topic_id: topic.id(),
+                                    bytes,
+                                })
+                                .await
+                        }
+                    };
+
+                    if let Err(err) = result {
+                        // @TODO(adz): This fails silently right now, shouldn't this be propagated
+                        // further to the user?
+                        error!("failed broadcasting message to gossip for topic {topic:?}: {err}");
+                        break;
+                    }
```

Comment:
And the reason this would fail is if there was an error sending on the gossip_actor_tx channel?

Reply:
Right, this should only happen if the receiver got dropped and we didn't close everything properly (which is the case right now), but it's otherwise fine?

Reply:
It doesn't propagate but would it be useful to handle this error? https://github.com/p2panda/p2panda/blob/net-cleanup/p2panda-net%2Fsrc%2Fengine%2Fgossip.rs#L107

It means that your message was not broadcasted which maybe is good to know?

Reply:
I'll add it to the roadmap :+1: